libsolidity/analysis/NameAndTypeResolver.cpp

```diff
@@ -703,6 +703,26 @@ void DeclarationRegistrationHelper::registerDeclaration(Declaration& _declaratio
 	solAssert(m_currentScope && m_scopes.count(m_currentScope), "No current scope.");
 	solAssert(m_currentScope == _declaration.scope(), "Unexpected current scope.");
 
+	// Functions defined inside quantifiers are available in the scope containing the quantifier
+	auto const* quantifier = dynamic_cast<ForAllQuantifier const*>(m_currentScope);
+	auto const* functionDefinition = dynamic_cast<FunctionDefinition const*>(&_declaration);
+	if (quantifier && functionDefinition)
+	{
+		solAssert(quantifier->scope());
+		solAssert(
+			// forall quantifiers cannot be used in block scope so the declaration is always active.
+			!dynamic_cast<Block const*>(quantifier->scope()) &&
+			!dynamic_cast<ForStatement const*>(quantifier->scope())
+		);
+		registerDeclaration(*m_scopes.at(quantifier->scope()), _declaration, nullptr, nullptr, false /* inactive */, m_errorReporter);
+
+		// TMP: Should I adjust the scope to match visibility? Here or in the Scoper?
+		//_declaration.annotation().scope = quantifier->scope();
+
+		solAssert(_declaration.annotation().contract == m_currentContract, "");
+		return;
+	}
```

Comment:
I hacked this in here for `ForAllQuantifier`, but seeing how other things related to scopes were done in #9586, I suspect we'd want to have a more general mechanism for things like this?

The problem I'm dealing with here is that `ForAllQuantifier` creates a new scope that contains its type variables and the function, which makes them invisible outside of it. For type variables it's exactly what we want but the function must be exposed to be callable.

One way to generalize this would be to make it possible to mark `Declaration`s as visible outside of their scope. I actually thought that `Declaration::isVisibleAsUnqualifiedName()` could already be used for that, but apparently it's just for hiding qualified things that are already in your scope.

Also, still not sure if the `scope` annotation not matching the scope where the declaration is registered can mess with things.

Reply:
I mean, this won't happen non-experimentally - and in the final version of experimental solidity, I don't see us still using the name and type resolver of legacy solidity - that's just a temporary hack to not have to bother with doing it properly for now. So since this hack seems to work and is very localized and very clearly only applies to experimental solidity, I'd say it's fine.